### PR TITLE
Merge fastCover into DictBuilder

### PR DESCRIPTION
**Pull request by jennifermliu**

### CODE REVIEW

File: lib/dictBuilder/zdict.h [hunk 1]
```diff
@@ -90,6 +91,17 @@ typedef struct {
...
+    unsigned d;                  /* dmer size : constraint: 0 < d <= k : Reasonable range [6, 16] */
+    unsigned f;                  /* log of size of frequency array : constraint: 0 < f <= 31 : 1 means default(18)*/
+    unsigned steps;              /* Number of steps : Only used for optimization : 0 means default (32) : Higher means more parameters checked */
+    unsigned nbThreads;          /* Number of threads : constraint: 0 < nbThreads : 1 means single-threaded : Only used for optimization : Ignored if ZSTD_MULTITHREAD is not defined */
+    double splitPoint;           /* Percentage of samples used for training: the first nbSamples * splitPoint samples will be used to training, the last nbSamples * (1 - splitPoint) samples will be used for testing, 0 means default (1.0), 1.0 when all samples are used for both training and testing */
+    unsigned finalize;           /* Percentage of training samples used for ZDICT_finalizeDictionary: 0 means default(100) */
+    unsigned skip;               /* Number of dmer skipped between each dmer counted in computeFrequency: constraint: 0 <= skip < k */
```
Comment (terrelln) [inline lib/dictBuilder/zdict.h]: This is getting to be a lot of parameters. Lets combine `skip` and `finalize` into a single parameter, called `skip`, or `acceleration`, etc.

We don't need to go into details of exactly what it does, just say higher means faster and less accurate, and 1 is the default (0 can clamp to 1). We'll use it internally to skip in `computeFrequency`, and decide on the number of samples passed to `ZSTD_finalizeDictionary()`.

File: lib/dictBuilder/zdict.h [hunk 2]
```diff
@@ -90,6 +91,17 @@ typedef struct {
     ZDICT_params_t zParams;
 } ZDICT_cover_params_t;
 
+typedef struct {
+    unsigned k;                  /* Segment size : constraint: 0 < k : Reasonable range [16, 2048+] */
+    unsigned d;                  /* dmer size : constraint: 0 < d <= k : Reasonable range [6, 16] */
+    unsigned f;                  /* log of size of frequency array : constraint: 0 < f <= 31 : 1 means default(18)*/
+    unsigned steps;              /* Number of steps : Only used for optimization : 0 means default (32) : Higher means more parameters checked */
+    unsigned nbThreads;          /* Number of threads : constraint: 0 < nbThreads : 1 means single-threaded : Only used for optimization : Ignored if ZSTD_MULTITHREAD is not defined */
+    double splitPoint;           /* Percentage of samples used for training: the first nbSamples * splitPoint samples will be used to training, the last nbSamples * (1 - splitPoint) samples will be used for testing, 0 means default (1.0), 1.0 when all samples are used for both training and testing */
```
Comment (terrelln) [inline lib/dictBuilder/zdict.h]: I think we shouldn't be using 1.0 as the default for fast cover. Evaluation is actually a significant part of the time that fast cover takes. We should select the highest value that works well enough from your previous experiments.

File: programs/zstd.1.md [hunk 3]
```diff
@@ -254,6 +254,21 @@ Compression of small files similar to the sample set will be greatly improved.
 
     `zstd --train-cover=k=50,split=60 FILEs`
 
+* `--train-fastcover[=k#,d=#,f=#,steps=#,split=#,accel=#]`:
+    Same as cover but with extra parameters f and accel and different default value of split
```
Comment (terrelln) [inline programs/zstd.1.md]: `finalize` and `skip` are out of date. Explain what `f` is.

Comment (jennifermliu) [reply programs/zstd.1.md]: Can you please explain what you mean by "finalize and skip are out of date"?

Comment (terrelln) [reply programs/zstd.1.md]: A few lines down you mention `finalize` and `skip`, but they don't exist anymore.

File: lib/dictBuilder/zdict.c [hunk 4]
```diff
@@ -1088,7 +1088,7 @@ size_t ZDICT_trainFromBuffer(void* dictBuffer, size_t dictBufferCapacity,
 #if defined(DEBUGLEVEL) && (DEBUGLEVEL>=1)
```
[the remark this answers is not in the document]
Comment (jennifermliu) [reply lib/dictBuilder/zdict.c]: I tried benchmarking with compressing at level 6 with dictionary generated at level 3 and compressing at level 6 with dictionary generated at level 6. The compression ratio of compressing at level 6 with dictionary generated at level 3 is 0.15-0.35 (depending on f and accel) below compressing at level 3 with dictionary generated at level 3 for github. The compression ratio of compressing at level 6 with dictionary generated at level 6 is ~0.1-0.2 higher than compressing at level 3 with dictionary generated at level 3, but with significantly longer time. It seems like generating dictionary at lower compression level and compressing at higher compression level hurts the compression ratio.

File: lib/BUCK [hunk 5]
```diff
@@ -66,6 +66,7 @@ cxx_library(
     visibility=['PUBLIC'],
     exported_headers=subdir_glob([
         ('dictBuilder', 'zdict.h'),
+        ('dictBuilder', 'cover.h'),
```
Comment (terrelln) [inline lib/BUCK]: This belongs in `headers`, not `exported_headers`.

File: lib/dictBuilder/zdict.h [hunk 6]
```diff
@@ -130,6 +141,48 @@ ZDICTLIB_API size_t ZDICT_optimizeTrainFromBuffer_cover(
...
+/*! ZDICT_trainFromBuffer_fastCover():
+ *  Train a dictionary from an array of samples using a modified version of COVER algorithm.
+ *  Samples must be stored concatenated in a single flat buffer `samplesBuffer`,
+ *  supplied with an array of sizes `samplesSizes`, providing the size of each sample, in order.
+ *  d and k are requiredMem.
+ *  All other parameters are optional, will use default values if not provided
+ *  The resulting dictionary will be saved into `dictBuffer`.
+ * @return: size of dictionary stored into `dictBuffer` (<= `dictBufferCapacity`)
+ *          or an error code, which can be tested with ZDICT_isError().
+ *  Note: ZDICT_trainFromBuffer_fastCover() requires about 9 bytes of memory for each input byte.
```
Comment (terrelln) [inline lib/dictBuilder/zdict.h]: This is no longer accurate, right?

File: lib/dictBuilder/zdict.h [hunk 7]
```diff
@@ -130,6 +141,48 @@ ZDICTLIB_API size_t ZDICT_optimizeTrainFromBuffer_cover(
     const void* samplesBuffer, const size_t* samplesSizes, unsigned nbSamples,
           ZDICT_cover_params_t* parameters);
 
+/*! ZDICT_trainFromBuffer_fastCover():
+ *  Train a dictionary from an array of samples using a modified version of COVER algorithm.
+ *  Samples must be stored concatenated in a single flat buffer `samplesBuffer`,
+ *  supplied with an array of sizes `samplesSizes`, providing the size of each sample, in order.
+ *  d and k are requiredMem.
```
Comment (terrelln) [inline lib/dictBuilder/zdict.h]: `required`, not `requiredMem`.

File: lib/dictBuilder/zdict.h [hunk 8]
```diff
@@ -130,6 +141,48 @@ ZDICTLIB_API size_t ZDICT_optimizeTrainFromBuffer_cover(
...
+                    const size_t *samplesSizes, unsigned nbSamples,
+                    ZDICT_fastCover_params_t parameters);
+
+/*! ZDICT_optimizeTrainFromBuffer_fastCover():
+ * The same requirements as above hold for all the parameters except `parameters`.
+ * This function tries many parameter combinations (specifically, k and d combinations)
+ * and picks the best parameters. `*parameters` is filled with the best parameters found,
+ * dictionary constructed with those parameters is stored in `dictBuffer`.
+ * All of the parameters d, k, steps, f, and accel are optional.
+ * If d is non-zero then we don't check multiple values of d, otherwise we check d = {6, 8, 10, 12, 14, 16}.
```
Comment (terrelln) [inline lib/dictBuilder/zdict.h]: This isn't correct anymore, right? We should only check 6, and 8. The docs for `ZDICT_optimizeTrainFromBuffer_cover()` should be fixed as well.

File: lib/dictBuilder/zdict.h [hunk 9]
```diff
@@ -130,6 +141,48 @@ ZDICTLIB_API size_t ZDICT_optimizeTrainFromBuffer_cover(
...
+ * if steps is zero it defaults to its default value.
+ * If k is non-zero then we don't check multiple values of k, otherwise we check steps values in [16, 2048].
+ * If f is zero, default value of 18 is used.
+ * If accel is zero, default value of 1 is used.
+ *
+ * @return: size of dictionary stored into `dictBuffer` (<= `dictBufferCapacity`)
+ *           or an error code, which can be tested with ZDICT_isError().
+ *           On success `*parameters` contains the parameters selected.
+ * Note: ZDICT_optimizeTrainFromBuffer_fastCover() requires about 8 bytes of memory for each input byte and additionally another 5 bytes of memory for each byte of memory for each thread.
```
Comment (terrelln) [inline lib/dictBuilder/zdict.h]: This isn't right anymore.

File: programs/dibio.h [hunk 10]
```diff
@@ -33,7 +33,7 @@
 */
 int DiB_trainFromFiles(const char* dictFileName, unsigned maxDictSize,
                        const char** fileNamesTable, unsigned nbFiles, size_t chunkSize,
-                       ZDICT_legacy_params_t* params, ZDICT_cover_params_t* coverParams,
-                       int optimizeCover);
+                       ZDICT_legacy_params_t *params, ZDICT_cover_params_t *coverParams,
```
Comment (terrelln) [inline programs/dibio.h]: nit: Switch back to `ZDICT_legacy_params_t* params` etc.

File: programs/dibio.c [hunk 11]
```diff
@@ -272,10 +272,11 @@ size_t ZDICT_trainFromBuffer_unsafe_legacy(void* dictBuffer, size_t dictBufferCa
 int DiB_trainFromFiles(const char* dictFileName, unsigned maxDictSize,
                        const char** fileNamesTable, unsigned nbFiles, size_t chunkSize,
                        ZDICT_legacy_params_t *params, ZDICT_cover_params_t *coverParams,
-                       int optimizeCover)
+                       ZDICT_fastCover_params_t *fastCoverParams, int optimize)
```
Comment (terrelln) [inline programs/dibio.c]: nit Switch to `ZDICT_fastCover_params_t* fastCoverParams`.

File: programs/zstd.1.md [hunk 12]
```diff
@@ -254,6 +254,26 @@ Compression of small files similar to the sample set will be greatly improved.
...
     `zstd --train-cover=k=50,split=60 FILEs`
 
+* `--train-fastcover[=k#,d=#,f=#,steps=#,split=#,accel=#]`:
+    Same as cover but with extra parameters _f_ and _accel_ and different default value of split
+    If _split_ is not specified, then it tries _split_ = 75.
+    If _f_ is not specified, then it tries _f_ = 18.
+    Requires that 0 < _f_ < 32.
+    If _accel_ is not specified, then it tries _accel_ = 1.
+    Requires that 0 < _accel_ <= 10.
+    Requires that _d_ = 6 _d_ = 8.
```
Comment (terrelln) [inline programs/zstd.1.md]: _d_ = 6 or _d_ = 8

File: lib/dictBuilder/fastcover.c [hunk 13]
```diff
@@ -0,0 +1,745 @@
...
+  /* Try each segment (activeSegment) and save the best (bestSegment) */
+  COVER_segment_t bestSegment = {0, 0, 0};
+  COVER_segment_t activeSegment;
+
+  /* Reset the activeDmers in the segment */
+  /* The activeSegment starts at the beginning of the epoch. */
+  activeSegment.begin = begin;
+  activeSegment.end = begin;
+  activeSegment.score = 0;
+  {
```
Comment (terrelln) [inline lib/dictBuilder/fastcover.c]: nit: This extra scope is unnecessary.

File: lib/dictBuilder/fastcover.c [hunk 14]
```diff
@@ -0,0 +1,745 @@
...
+
+
+/**
+ * Clean up a context initialized with `FASTCOVER_ctx_init()`.
+ */
+static void FASTCOVER_ctx_destroy(FASTCOVER_ctx_t *ctx) {
+  if (!ctx) {
+    return;
+  }
+  if (ctx->segmentFreqs) {
```
Comment (terrelln) [inline lib/dictBuilder/fastcover.c]: nit: We can remove these if statments, `free()` handles `NULL` arguments.

```c
free(ctx->segmentFreqs);
ctx->segmentFreqs = NULL;
```

File: lib/dictBuilder/fastcover.c [hunk 15]
```diff
@@ -0,0 +1,745 @@
...
+    if (dst) {
+      free(dst);
+    }
+  }
+
+_cleanup:
+  COVER_best_finish(data->best, totalCompressedSize, parameters, dict,
+                    dictBufferCapacity);
+  free(data);
+  if (dict) {
```
Comment (terrelln) [inline lib/dictBuilder/fastcover.c]: nit: No need for the if statements like above.

File: lib/dictBuilder/fastcover.c [hunk 16]
```diff
@@ -0,0 +1,745 @@
...
+    dictBufferCapacity = ZDICT_finalizeDictionary(
+        dict, dictBufferCapacity, dict + tail, dictBufferCapacity - tail,
+        ctx->samples, ctx->samplesSizes, nbFinalizeSamples, parameters.zParams);
+    if (ZDICT_isError(dictBufferCapacity)) {
+      DISPLAYLEVEL(1, "Failed to finalize dictionary\n");
+      goto _cleanup;
+    }
+  }
+  /* Check total compressed size */
+  {
```
Comment (terrelln) [inline lib/dictBuilder/fastcover.c]: Lets make the total compressed size sub-routine a function that is shared between cover and fast-cover.

File: lib/dictBuilder/fastcover.c [hunk 17]
```diff
@@ -0,0 +1,745 @@
...
+    if (accel == 0 || accel > FASTCOVER_MAX_ACCEL) {
+      LOCALDISPLAYLEVEL(displayLevel, 1, "Incorrect accel\n");
+      return ERROR(GENERIC);
+    }
+    if (kMinK < kMaxD || kMaxK < kMinK) {
+      LOCALDISPLAYLEVEL(displayLevel, 1, "Incorrect k\n");
+      return ERROR(GENERIC);
+    }
+    if (nbSamples == 0) {
+      DISPLAYLEVEL(1, "FASTCOVER must have at least one input file\n");
```
Comment (terrelln) [inline lib/dictBuilder/fastcover.c]: `LOCALDISPLAYLEVEL()`

File: lib/dictBuilder/fastcover.c [hunk 18]
```diff
@@ -0,0 +1,745 @@
...
+      return ERROR(GENERIC);
+    }
+    if (dictBufferCapacity < ZDICT_DICTSIZE_MIN) {
+      DISPLAYLEVEL(1, "dictBufferCapacity must be at least %u\n",
```
Comment (terrelln) [inline lib/dictBuilder/fastcover.c]: `LOCALDISPLAYLEVEL()`.

File: tests/playTests.sh [hunk 19]
```diff
@@ -444,6 +444,45 @@ $ZSTD --train-cover *.c ../programs/*.c
...
+$ZSTD --train-fastcover=f=16 -r *.c ../programs/*.c
+$ECHO "- Create dictionary using accel=2"
+$ZSTD --train-fastcover=accel=2 -r *.c ../programs/*.c
+$ECHO "- Create dictionary using accel=10"
+$ZSTD --train-fastcover=accel=10 -r *.c ../programs/*.c
+$ECHO "- Test -o before --train-fastcover"
+rm -f tmpDict dictionary
+$ZSTD -o tmpDict --train-fastcover *.c ../programs/*.c
+test -f tmpDict
+$ZSTD --train-fastcover *.c ../programs/*.c
```
Comment (terrelln) [inline tests/playTests.sh]: You should add some test cases with multithreading enabled (`-T4`).

Comment (terrelln) [reply tests/playTests.sh]: We run `playTests.sh` with thread sanitizer (TSAN), so if we have a multithreaded test, we should catch some of the multithreading bugs in CI.

File: lib/dictBuilder/fastcover.c [hunk 20]
```diff
@@ -148,6 +147,9 @@ static COVER_segment_t FASTCOVER_selectSegment(const FASTCOVER_ctx_t *ctx,
   COVER_segment_t bestSegment = {0, 0, 0};
   COVER_segment_t activeSegment;
 
+  /* Initialize array to keep track of frequency of dmer within activeSegment */
+  U16* segmentFreqs = (U16 *)calloc(((U64)1 << f), sizeof(U16));
```
Comment (terrelln) [inline lib/dictBuilder/fastcover.c]: This doesn't belong in selectSegment, we only need one copy per thread, but we don't need to allocate/free it in the hot loop. This probably belongs in `FASTCOVER_buildDictionary()`.

**Pull request merged**